### PR TITLE
[Bounty] Update Dr. Agent model (PyHealth 2.0)

### PR DESCRIPTION
**Pull request by joshuasteier**
## Contributor
- **Name:** Joshua Steier
## Contribution Type
Model

## Description
Implements the Dr. Agent model from "Dr. Agent: Clinical predictive model via mimicked second opinions" (JAMIA 2020).

**Key features:**
- Dual-agent reinforcement learning with dynamic skip connections
- Primary agent (current visit) + second-opinion agent (historical context)
- Support for static demographic features
- GRU/LSTM cell options
- REINFORCE with optional baseline for variance reduction

**Paper:** https://doi.org/10.1093/jamia/ocaa074

## Files to Review
| File | Description |
|------|-------------|
| `pyhealth/models/agent.py` | Model implementation (AgentLayer + Agent) |
| `tests/test_agent.py` | Unit tests |
| `examples/agent_mimic4.ipynb` | Example notebook with MIMIC-IV |

## How to Test
```bash
# Run unit tests
python -m pytest tests/test_agent.py -v

# Or run the model directly
python pyhealth/models/agent.py
```

## Checklist
- [x] Code follows PEP8 (88-char line limit)
- [x] Google-style docstrings with type hints
- [x] Inherits from BaseModel (v2.0 style)
- [x] Unit tests provided
- [x] Example notebook provided
- [x] Rebased with main branch

### CODE REVIEW

File: tests/core/test_agent.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+            },
+            {
+                "patient_id": "patient-1",
+                "visit_id": "visit-1",
+                "conditions": ["B01", "B02", "B03"],
+                "procedures": ["P4", "P5"],
+                "demographic": [45.0, 0.0, 22.1],
+                "label": 0,
+            },
+        ]
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply tests/core/test_agent.py:37]: Fixed. Added torch.sum(x, dim=2) for NestedSequenceProcessor outputs

File: pyhealth/models/agent.py [hunk 2]
```diff
@@ -189,475 +219,410 @@ def forward(
...
+        # Get embeddings for sequence features
+        embedded = self.embedding_model(kwargs)
+
+        # Process each sequence feature through its agent
+        for feature_key in self.seq_feature_keys:
+            x = embedded[feature_key]
+            if not isinstance(x, torch.Tensor):
+                x = torch.tensor(x, dtype=torch.float, device=self.device)
+            x = x.float().to(self.device)
+
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/agent.py]: Fixed. Removed unnecessary type checking

File: pyhealth/models/agent.py [hunk 3]
```diff
@@ -189,475 +219,410 @@ def forward(
...
+            # Compute mask from embeddings (non-zero entries are valid)
+            mask = x.sum(dim=-1) != 0
+            mask_dict[feature_key] = mask
+
+            # Forward through agent layer
+            out, _ = self.agent[feature_key](x, static=static, mask=mask)
+            patient_emb.append(out)
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/agent.py:613]: Fixed. Added torch.sum(x, dim=2) for NestedSequenceProcessor outputs.

File: pyhealth/models/agent.py [hunk 4]
```diff
@@ -189,475 +219,410 @@ def forward(
...
+        ...         "visit_id": "v1",
+        ...         "conditions": [["C01"]],
+        ...         "procedures": [["P4"]],
+        ...         "demographic": [45.0, 0.0, 22.1],
+        ...         "label": 0,
+        ...     },
+        ... ]
+        >>> dataset = SampleDataset(
+        ...     samples=samples,
+        ...     input_schema={"conditions": "sequence", "procedures": "sequence"},
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/agent.py]: Fixed, uses nested_sequence schema.

File: tests/core/test_agent.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+        train_loader = get_dataloader(dataset, batch_size=3, shuffle=False)
+        data_batch = next(iter(train_loader))
+
+        with torch.no_grad():
+            ret = model(**data_batch)
+
+        self.assertEqual(ret["y_prob"].shape, (3, 3))
+        self.assertIn("loss", ret)
+
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply tests/core/test_agent.py:241]: Fixed. Added shape and softmax assertions

File: pyhealth/models/agent.py [hunk 6]
```diff
@@ -189,475 +219,410 @@ def forward(
...
+
+        Returns:
+            Combined RL loss (policy loss + value loss if using baseline).
+        """
+        # Compute rewards based on prediction accuracy
         if self.mode == "binary":
-            pred = torch.sigmoid(pred)
-            rewards = ((pred - 0.5) * 2 * true).squeeze()
+            pred_prob = torch.sigmoid(pred)
+            rewards = ((pred_prob - 0.5) * 2 * true).squeeze()
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/agent.py]: Fixed. Changed to (1 - torch.abs(true - pred_prob))

[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/agent.py]: @jhnwu3 @LogicFan  I believe all the errors are resolved in the latest commit.

**Review by LogicFan — APPROVED** [review]
